core/src/main/java/edu/wpi/grip/core/PipelineRunner.java

```diff
@@ -0,0 +1,159 @@
+import edu.wpi.grip.core.util.service.RestartableService;
+
+import javax.inject.Inject;
+import java.util.concurrent.CountDownLatch;
+import java.util.concurrent.TimeUnit;
+import java.util.function.Supplier;
+import java.util.logging.Logger;
+
+/**
+ * Runs the pipeline in a separate thread.
+ * The runner listens for
+ */
+@Singleton
+public class PipelineRunner implements StartStoppable {
+    private final Logger logger = Logger.getLogger(getClass().getName());
+    /**
+     * This is used to flag that the pipeline needs to run because of an update
+     */
+    private final CountDownLatch pipelineFlag = new CountDownLatch(1);
+    private final Provider<Pipeline> pipelineProvider;
+    private final RestartableService pipelineService;
+
+    @Inject
+    PipelineRunner(EventBus eventBus, Provider<Pipeline> pipelineProvider) {
+        this.pipelineProvider = pipelineProvider;
+        this.pipelineService = new RestartableService(
+                () -> new AbstractScheduledService(){
+
+                    @Override
+                    protected void startUp() {
+                        logger.info("Pipeline Starting");
+                    }
+
+                    @Override
+                    protected void runOneIteration() {
+                        if (!isRunning()) return;
+                        try {
+                            pipelineFlag.await();
+                        } catch (InterruptedException e) {
+                            Thread.currentThread().interrupt();
+                            return;
+                        }
+                        if (!isRunning()) return;
+                        runPipeline(this::isRunning);
+                        // This should not block access to the steps array
+                        if (isRunning()) {
+                            eventBus.post(new RenderEvent());
+                        }
+                    }
+
+                    @Override
+                    protected void shutDown() {
+                        logger.info("Pipeline Shutting Down");
+                    }
+
+                    @Override
+                    protected Scheduler scheduler() {
+                        return Scheduler.newFixedRateSchedule(0, 1, TimeUnit.MILLISECONDS);
+                    }
+                }
+        );
+        this.pipelineService.addListener(
+                new EventBusServiceListener(eventBus::post, this),
+                MoreExecutors.directExecutor());
+    }
+
+    /**
+     * Starts the pipeline to run at the default rate.
+     */
+    @Override
+    public void startAsync() {
+        pipelineService.startAsync();
+
+        pipelineFlag.countDown();
+    }
+
+    @Override
+    public void stopAsync() {
+        pipelineService.stopAsync();
+        // Ensure that we unblock the pipeline so it can actually stop
+        pipelineFlag.countDown();
+    }
+
+    @Override
+    public Service.State getState() {
+        return pipelineService.state();
```

Comment:
I'm not pulling all of the methods out of service.
I'm only using 4/10 of them.
I don't need the other 6.